File: ios/TextRecognitionPlugin.swift
//
//  TextRecognitionPlugin.swift
//  react-native-vision-camera-ml-kit
//

import Foundation
import VisionCamera
import MLKitVision
import MLKitCommon
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean

@objc(TextRecognitionPlugin)
public class TextRecognitionPlugin: FrameProcessorPlugin {

    private var textRecognizer: TextRecognizer!
    private var textLayout: TextLayout = .horizontal
    private var async = false
    private lazy var stackedReader = StackedTextReader()
    private let stackedQueue = DispatchQueue(label: "stacked-text", qos: .userInitiated)
    private let processingQueue = DispatchQueue(label: "text-recognition", qos: .userInitiated)
    private let luma = ImageUtils.LumaPlane()
    private let lock = NSLock()
    private var processing = false
    private var pendingResult: [String: Any]?

    private struct Captured {
        let image: UIImage
        let orientation: UIImage.Orientation
        let degrees: Int
        let stacked: Bool
        let startTime: Date
    }

    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        super.init(proxy: proxy, options: options)

        let language = (options["language"] as? String ?? "latin").lowercased()
        Logger.info("Initializing text recognition with language: \(language)")

        switch language {
        case "chinese":
            textRecognizer = TextRecognizer.textRecognizer(options: ChineseTextRecognizerOptions())
        case "devanagari":
            textRecognizer = TextRecognizer.textRecognizer(options: DevanagariTextRecognizerOptions())
        case "japanese":
            textRecognizer = TextRecognizer.textRecognizer(options: JapaneseTextRecognizerOptions())
        case "korean":
            textRecognizer = TextRecognizer.textRecognizer(options: KoreanTextRecognizerOptions())
        case "latin", "default":
            textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
        default:
            Logger.warn("Unknown language '\(language)', defaulting to Latin")
            textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
        }

        textLayout = TextLayout.from(options["textLayout"])
        async = options["async"] as? Bool ?? false
        Logger.info("Text recognition initialized successfully (textLayout: \(textLayout.rawValue), async: \(async))")
    }

    deinit {
        Logger.debug("TextRecognitionPlugin deallocating - ML Kit recognizer resources will be freed")
    }

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        let startTime = Date()

        if !async {
            guard let captured = capture(frame, startTime: startTime) else { return nil }
            return process(captured)
        }

        lock.lock()
        if processing {
            lock.unlock()
            return nil
        }
        let finished = pendingResult
        pendingResult = nil
        lock.unlock()

        guard let captured = capture(frame, startTime: startTime) else { return finished }
        lock.lock()
        processing = true
        lock.unlock()
        processingQueue.async { [self] in
            let result = process(captured)
            lock.lock()
            pendingResult = result
            processing = false
            lock.unlock()
        }
        return finished
    }

    private func capture(_ frame: Frame, startTime: Date) -> Captured? {
        let orientation = frame.orientation

        guard let clonedImage = ImageUtils.imageFromSampleBuffer(frame.buffer) else {
            Logger.error("Failed to create vision image from sample buffer")
            return nil
        }
        let imageOrientation = getOrientation(orientation: orientation)

        Logger.debug("Processing frame: \(frame.width)x\(frame.height), orientation: \(orientation.rawValue)")

        let degrees = TextRecognitionPlugin.rotationDegrees(for: imageOrientation)
        var stacked = false
        if textLayout != .horizontal, let pixelBuffer = CMSampleBufferGetImageBuffer(frame.buffer) {
            stacked = ImageUtils.readLuma(pixelBuffer, rotationDegrees: degrees, longSide: StackedTextReader.frameDetectionLongSide, into: luma)
        }
        return Captured(image: clonedImage, orientation: imageOrientation, degrees: degrees, stacked: stacked, startTime: startTime)
    }

    private func process(_ captured: Captured) -> [String: Any]? {
        let clonedImage = captured.image
        let degrees = captured.degrees
        let visionImage = VisionImage(image: clonedImage)
        visionImage.orientation = captured.orientation

        let uprightSize = degrees % 180 == 0 ? clonedImage.size : CGSize(width: clonedImage.size.height, height: clonedImage.size.width)
        var prepared = [StackedTextReader.Prepared]()
        let group = DispatchGroup()
        var pending = false
        if textLayout == .stacked, captured.stacked, let source = clonedImage.cgImage {
            pending = true
            group.enter()
            stackedQueue.async { [self] in
                prepared = stackedReader.prepare(luma: luma, source: source, rotationDegrees: degrees)
                group.leave()
            }
        }

        do {
            let text: Text
            do {
                text = try textRecognizer.results(in: visionImage)
            } catch {
                if pending { group.wait() }
                throw error
            }

            let stackedBlocks: [StackedTextReader.StackedBlock]
            if pending {
                group.wait()
                stackedBlocks = stackedReader
                    .recognize(prepared, recognizer: textRecognizer)
                    .map { StackedTextReader.toSourceCoordinates($0, degrees: degrees, rotatedSize: uprightSize) }
            } else if captured.stacked, textLayout.shouldReadStacked(text), let source = clonedImage.cgImage {
                stackedBlocks = stackedReader
                    .recognize(stackedReader.prepare(luma: luma, source: source, rotationDegrees: degrees), recognizer: textRecognizer)
                    .map { StackedTextReader.toSourceCoordinates($0, degrees: degrees, rotatedSize: uprightSize) }
            } else {
                stackedBlocks = []
            }

            let processingTime = Int64(Date().timeIntervalSince(captured.startTime) * 1000)
            Logger.performance("Text recognition processing", durationMs: processingTime)

            if text.text.isEmpty && stackedBlocks.isEmpty {
                Logger.debug("No text detected in frame")
                return nil
            }

            Logger.debug("Text detected: \(text.text.count) characters, \(text.blocks.count) blocks, \(stackedBlocks.count) stacked")

            var blocks = processBlocks(text.blocks)
            blocks.append(contentsOf: stackedBlocks.map { StackedTextBlocks.toDictionary($0) })

            return [
                "text": StackedTextBlocks.combineText(text.text, stackedBlocks),
                "blocks": blocks
            ]
        } catch {
            let processingTime = Int64(Date().timeIntervalSince(captured.startTime) * 1000)
            Logger.error("Error during text recognition: \(error.localizedDescription)")
            Logger.performance("Text recognition processing (error)", durationMs: processingTime)
            return nil
        }
    }

    // MARK: - Orientation Mapping

    private func getOrientation(orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
        case .up:
            return .up
        case .left:
            return .right  // Swap left and right
        case .down:
            return .down
        case .right:
            return .left   // Swap left and right
        default:
            return .up
        }
    }

    private static func rotationDegrees(for orientation: UIImage.Orientation) -> Int {
        switch orientation {
        case .right: return 90
        case .down: return 180
        case .left: return 270
        default: return 0
        }
    }

    // MARK: - Helper Methods

    private func processBlocks(_ blocks: [TextBlock]) -> [[String: Any]] {
        return blocks.map { block in
            var blockDict: [String: Any] = [
                "text": block.text,
                "frame": processRect(block.frame),
                "cornerPoints": processCornerPoints(block.cornerPoints),
                "lines": processLines(block.lines)
            ]

            if let lang = block.recognizedLanguages.first?.languageCode {
                blockDict["recognizedLanguage"] = lang
            }

            return blockDict
        }
    }

    private func processLines(_ lines: [TextLine]) -> [[String: Any]] {
        return lines.map { line in
            var lineDict: [String: Any] = [
                "text": line.text,
                "frame": processRect(line.frame),
                "cornerPoints": processCornerPoints(line.cornerPoints),
                "elements": processElements(line.elements)
            ]

            if let lang = line.recognizedLanguages.first?.languageCode {
                lineDict["recognizedLanguage"] = lang
            }

            return lineDict
        }
    }

    private func processElements(_ elements: [TextElement]) -> [[String: Any]] {
        return elements.map { element in
            var elementDict: [String: Any] = [
                "text": element.text,
                "frame": processRect(element.frame),
                "cornerPoints": processCornerPoints(element.cornerPoints)
            ]

            if let lang = element.recognizedLanguages.first?.languageCode {
                elementDict["recognizedLanguage"] = lang
            }

            return elementDict
        }
    }

    private func processRect(_ rect: CGRect) -> [String: CGFloat] {
        return [
            "x": rect.midX,
            "y": rect.midY,
            "width": rect.width,
            "height": rect.height
        ]
    }

    private func processCornerPoints(_ cornerPoints: [NSValue]) -> [[String: Int]] {
        return cornerPoints.compactMap { $0.cgPointValue }.map { point in
            ["x": Int(point.x), "y": Int(point.y)]
        }
    }
}

File: android/src/main/java/com/rnvisioncameramlkit/TextRecognitionPlugin.kt
package com.rnvisioncameramlkit

import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.Rect
import android.media.Image
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import com.rnvisioncameramlkit.stacked.StackedTextBlocks
import com.rnvisioncameramlkit.stacked.StackedTextReader
import com.rnvisioncameramlkit.stacked.TextLayout
import com.rnvisioncameramlkit.utils.ImageUtils
import com.rnvisioncameramlkit.utils.Logger
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future

class TextRecognitionPlugin(
    proxy: VisionCameraProxy,
    options: Map<String, Any>?
) : FrameProcessorPlugin(), AutoCloseable {

    private var recognizer: TextRecognizer
    private val textLayout: TextLayout
    private val async: Boolean
    private val stackedReader: StackedTextReader by lazy { StackedTextReader() }
    private val stackedExecutorDelegate = lazy {
        Executors.newSingleThreadExecutor { runnable -> Thread(runnable, "stacked-text") }
    }
    private val stackedExecutor: ExecutorService by stackedExecutorDelegate
    private val processingExecutorDelegate = lazy {
        Executors.newSingleThreadExecutor { runnable -> Thread(runnable, "text-recognition") }
    }
    private val processingExecutor: ExecutorService by processingExecutorDelegate
    private val luma = ImageUtils.LumaPlane()
    private var pendingResult: Future<HashMap<String, Any?>?>? = null

    private class Captured(val bitmap: Bitmap, val rotationDegrees: Int, val stacked: Boolean, val startTime: Long)

    init {
        val language = options?.get("language")?.toString() ?: "latin"
        Logger.info("Initializing text recognition with language: $language")

        val recognizerOptions: TextRecognizerOptionsInterface = when (language.lowercase()) {
            "chinese" -> ChineseTextRecognizerOptions.Builder().build()
            "devanagari" -> DevanagariTextRecognizerOptions.Builder().build()
            "japanese" -> JapaneseTextRecognizerOptions.Builder().build()
            "korean" -> KoreanTextRecognizerOptions.Builder().build()
            "latin", "default" -> TextRecognizerOptions.Builder().build()
            else -> {
                Logger.warn("Unknown language '$language', defaulting to Latin")
                TextRecognizerOptions.Builder().build()
            }
        }

        recognizer = TextRecognition.getClient(recognizerOptions)
        textLayout = TextLayout.from(options?.get("textLayout")?.toString())
        async = options?.get("async") == true
        Logger.info("Text recognition initialized successfully (textLayout: $textLayout, async: $async)")
    }

    override fun close() {
        try {
            ImageUtils.clearBuffers()
            if (processingExecutorDelegate.isInitialized()) processingExecutor.shutdown()
            if (stackedExecutorDelegate.isInitialized()) stackedExecutor.shutdown()
            recognizer.close()
            Logger.debug("Text recognizer resources cleaned up successfully")
        } catch (e: Exception) {
            Logger.error("Error cleaning up text recognizer resources", e)
        }
    }

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
        val startTime = System.currentTimeMillis()
        try {
            if (!async) {
                val captured = capture(frame, startTime) ?: return null
                return process(captured)
            }

            val pending = pendingResult
            if (pending != null && !pending.isDone) return null
            pendingResult = null
            val finished = pending?.let { takeResult(it) }

            val captured = capture(frame, startTime) ?: return finished
            pendingResult = processingExecutor.submit(Callable { process(captured) })
            return finished
        } catch (e: Exception) {
            Logger.error("Error during text recognition", e)
            Logger.performance("Text recognition processing (error)", System.currentTimeMillis() - startTime)
            return null
        }
    }

    private fun takeResult(pending: Future<HashMap<String, Any?>?>): HashMap<String, Any?>? =
        try {
            pending.get()
        } catch (e: Exception) {
            Logger.error("Error during text recognition", e)
            null
        }

    private fun capture(frame: Frame, startTime: Long): Captured? {
        val mediaImage: Image = frame.image
        val rotationDegrees = frame.imageProxy.imageInfo.rotationDegrees

        if (Logger.isDebugEnabled()) {
            Logger.debug("Processing frame: ${frame.width}x${frame.height}, rotation: $rotationDegrees")
        }

        val clonedBitmap = ImageUtils.imageToBitmap(mediaImage, 0)
        if (clonedBitmap == null) {
            Logger.error("Failed to clone camera image to bitmap")
            return null
        }

        val stacked = textLayout != TextLayout.HORIZONTAL
        if (stacked) {
            try {
                ImageUtils.readLuma(mediaImage, rotationDegrees, StackedTextReader.FRAME_DETECTION_LONG_SIDE, luma)
            } catch (e: Exception) {
                clonedBitmap.recycle()
                throw e
            }
        }
        return Captured(clonedBitmap, rotationDegrees, stacked, startTime)
    }

    private fun process(captured: Captured): HashMap<String, Any?>? {
        val clonedBitmap = captured.bitmap
        val rotationDegrees = captured.rotationDegrees
        var pending: Future<List<StackedTextReader.Prepared>>? = null
        try {
            if (textLayout == TextLayout.STACKED) {
                pending = stackedExecutor.submit(Callable { stackedReader.prepare(luma, clonedBitmap, rotationDegrees) })
            }

            val image = InputImage.fromBitmap(clonedBitmap, rotationDegrees)
            val task: Task<Text> = recognizer.process(image)
            val text: Text = Tasks.await(task)

            val stackedBlocks = when {
                pending != null -> {
                    val prepared = awaitPrepared(pending)
                    pending = null
                    stackedReader.recognize(prepared, recognizer, STACKED_TIMEOUT_SECONDS)
                }
                captured.stacked && textLayout.shouldReadStacked(text) -> readStackedColumns(clonedBitmap, rotationDegrees)
                else -> emptyList()
            }

            Logger.performance("Text recognition processing", System.currentTimeMillis() - captured.startTime)

            if (text.text.isEmpty() && stackedBlocks.isEmpty()) {
                Logger.debug("No text detected in frame")
                return null
            }

            Logger.debug("Text detected: ${text.text.length} characters, ${text.textBlocks.size} blocks, ${stackedBlocks.size} stacked")

            val blocks = processBlocks(text.textBlocks)
            stackedBlocks.forEach { blocks.pushMap(StackedTextBlocks.toMap(it)) }

            val result = WritableNativeMap().apply {
                putString("text", StackedTextBlocks.combineText(text.text, stackedBlocks))
                putArray("blocks", blocks)
            }
            return result.toHashMap()
        } catch (e: Exception) {
            Logger.error("Error during text recognition", e)
            Logger.performance("Text recognition processing (error)", System.currentTimeMillis() - captured.startTime)
            return null
        } finally {
            pending?.let { awaitPrepared(it).forEach { prepared -> prepared.strip.recycle() } }
            clonedBitmap.recycle()
        }
    }

    private fun awaitPrepared(pending: Future<List<StackedTextReader.Prepared>>): List<StackedTextReader.Prepared> =
        try {
            pending.get()
        } catch (e: Exception) {
            Logger.error("Stacked text pass failed", e)
            emptyList()
        }

    private fun readStackedColumns(clonedBitmap: Bitmap, rotationDegrees: Int): List<StackedTextReader.StackedBlock> {
        return try {
            val prepared = stackedReader.prepare(luma, clonedBitmap, rotationDegrees)
            stackedReader.recognize(prepared, recognizer, STACKED_TIMEOUT_SECONDS)
        } catch (e: Exception) {
            Logger.error("Stacked text pass failed", e)
            emptyList()
        }
    }

    companion object {
        private const val STACKED_TIMEOUT_SECONDS = 5L

        /**
         * Process text blocks into React Native compatible format
         */
        private fun processBlocks(blocks: List<Text.TextBlock>): WritableNativeArray {
            val blockArray = WritableNativeArray()

            for (block in blocks) {
                val blockMap = WritableNativeMap().apply {
                    putString("text", block.text)
                    putMap("frame", processRect(block.boundingBox))
                    putArray("cornerPoints", processCornerPoints(block.cornerPoints))
                    putArray("lines", processLines(block.lines))

                    // Add language if recognized
                    block.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }

                    // Add confidence if available (ML Kit doesn't provide this for v2, but keeping for future)
                    // putDouble("confidence", block.confidence?.toDouble() ?: 0.0)
                }
                blockArray.pushMap(blockMap)
            }

            return blockArray
        }

        /**
         * Process text lines into React Native compatible format
         */
        private fun processLines(lines: List<Text.Line>): WritableNativeArray {
            val lineArray = WritableNativeArray()

            for (line in lines) {
                val lineMap = WritableNativeMap().apply {
                    putString("text", line.text)
                    putMap("frame", processRect(line.boundingBox))
                    putArray("cornerPoints", processCornerPoints(line.cornerPoints))
                    putArray("elements", processElements(line.elements))

                    line.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                lineArray.pushMap(lineMap)
            }

            return lineArray
        }

        /**
         * Process text elements (words) into React Native compatible format
         */
        private fun processElements(elements: List<Text.Element>): WritableNativeArray {
            val elementArray = WritableNativeArray()

            for (element in elements) {
                val elementMap = WritableNativeMap().apply {
                    putString("text", element.text)
                    putMap("frame", processRect(element.boundingBox))
                    putArray("cornerPoints", processCornerPoints(element.cornerPoints))
                    putArray("symbols", processSymbols(element.symbols))

                    element.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                elementArray.pushMap(elementMap)
            }

            return elementArray
        }

        /**
         * Process text symbols (characters) into React Native compatible format
         */
        private fun processSymbols(symbols: List<Text.Symbol>): WritableNativeArray {
            val symbolArray = WritableNativeArray()

            for (symbol in symbols) {
                val symbolMap = WritableNativeMap().apply {
                    putString("text", symbol.text)
                    putMap("frame", processRect(symbol.boundingBox))
                    putArray("cornerPoints", processCornerPoints(symbol.cornerPoints))

                    symbol.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                symbolArray.pushMap(symbolMap)
            }

            return symbolArray
        }

        /**
         * Convert Android Rect to React Native format
         */
        private fun processRect(boundingBox: Rect?): WritableNativeMap {
            val rectMap = WritableNativeMap()

            boundingBox?.let { box ->
                rectMap.putDouble("x", box.exactCenterX().toDouble())
                rectMap.putDouble("y", box.exactCenterY().toDouble())
                rectMap.putInt("width", box.width())
                rectMap.putInt("height", box.height())
            }

            return rectMap
        }

        /**
         * Convert Android corner points to React Native format
         */
        private fun processCornerPoints(cornerPoints: Array<Point>?): WritableNativeArray {
            val pointsArray = WritableNativeArray()

            cornerPoints?.forEach { point ->
                val pointMap = WritableNativeMap().apply {
                    putInt("x", point.x)
                    putInt("y", point.y)
                }
                pointsArray.pushMap(pointMap)
            }

            return pointsArray
        }
    }
}
